### PR TITLE
[v2] Refactor mieb to dataloaders

### PR DESCRIPTION
**Pull request by Samoed**
Refactor mieb to use dataloaders.
Closes https://github.com/embeddings-benchmark/mteb/issues/1606

Refactored and tested model wrappers:
- [x] align models 
- [x] blip2
- [x] blip
- [X] clip
- [x] cohere v
- [x] dino
- [x] eva clip
- [x] e5_v
- [x] gme
- [x] jina clip
- [x] llm2clip
- [x] vlm2vec
- [x] vista
- [x] voyage
- [x] siglip
- [x] openclip
- [x] nomic vision
- [x] moco

### Code Quality

- [X] **Code Formatted**: Format the code using `make lint` to maintain consistent style.

### Documentation

- [ ] **Updated Documentation**: Add or update documentation to reflect the changes introduced in this PR.

### Testing

- [ ] **New Tests Added**: Write tests to cover new functionality. Validate with `make test-with-coverage`.
- [ ] **Tests Passed**: Run tests locally using `make test` or `make test-with-coverage` to ensure no existing functionality is broken.

### Adding datasets checklist

**Reason for dataset addition**: ... 

- [ ] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [ ] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [ ] `intfloat/multilingual-e5-small`
- [ ] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [ ] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [ ] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [ ] Run tests locally to make sure nothing is broken using `make test`. 
- [ ] Run the formatter to format the code using `make lint`. 

### Adding a model checklist

 - [ ] I have filled out the ModelMeta object to the extent possible
 - [ ] I have ensured that my model can be loaded using
   - [ ] `mteb.get_model(model_name, revision)` and
   - [ ] `mteb.get_model_meta(model_name, revision)`
 - [ ] I have tested the implementation works on a representative set of tasks.

### CODE REVIEW

**Comment by isaac-chung** [comment]
Due to the long list of models, I'd suggest getting one ready, go through review cycle, then do the rest based on the 1st one.

**Comment by Samoed** [comment]
Yes, I'm testing all tasks on clip model. Only retrieval and multiple choice tasks are left to refactor

**Comment by Samoed** [comment]
I don't know how to refactor `ImageTextPairClassificationEvaluator`, because I tried multiple implementations and they are slow or required a lot of memory
<details>
<summary>My test implementation</summary>

```python
from __future__ import annotations

import logging
from typing import Any

import torch
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from datasets import Dataset
from mteb.create_dataloaders import prepare_image_dataset
from mteb.encoder_interface import Encoder, EncoderWithSimilarity
from mteb.evaluation.evaluators.Evaluator import Evaluator
from tqdm.auto import tqdm

logger = logging.getLogger(__name__)

def combine_columns(
    row: dict[str, Any],
    text_columns: list[str],
    image_columns: list[str],
) -> dict[str, list[Any]]:
    images = [row[col] for col in image_columns]
    texts = [row[col] for col in text_columns]
    return {
        "image": images,
        "text": texts,
    }


class ImageTextPairClassificationEvaluator(Evaluator):
    """Evaluate a model based on the similarity of the embeddings by calculating the accuracy of
    identifying similar and dissimilar image caption pairs.
    The goal is to find the correct image for each caption and the correct caption for each image.
    This is done by computing the similarities between each image and each caption.
    The results are written in a CSV. If a CSV already exists, then values are appended.
    The labels need to be 0 for dissimilar pairs and 1 for similar pairs.

    Args:
        images: Each row is a list of images.
        texts: Each row is a list of captions.
        batch_size: Batch size used to compute embeddings
    """

    def __init__(
        self,
        dataset,
        images_column_names: str | list[str],
        texts_column_names: str | list[str],
        task_name: str | None = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.dataset = dataset
        self.images_column_names = images_column_names
        self.texts_column_names = texts_column_names
        self.task_name = task_name

    def __call__(
        self,
        model: Encoder | EncoderWithSimilarity,
        encode_kwargs: dict[str, Any] = {},
    ):
        if "batch_size" not in encode_kwargs:
            encode_kwargs["batch_size"] = 64

        if isinstance(self.images_column_names, list) and isinstance(self.texts_column_names, list):
            # from collections import defaultdict
            # dataset = defaultdict(list)
            # for row in tqdm(self.dataset.iter(batch_size=1), total=len(self.dataset)):
            #     images = [row[col] for col in self.images_column_names]
            #     texts = [row[col] for col in self.texts_column_names]
            #     dataset["image"].append(images)
            #     dataset["text"].append(texts)
            #
            #     # dataset.append({
            #     #     "image": images,
            #     #     "text": texts,
            #     # })
            # dataset = Dataset.from_dict(dataset)
            dataset = self.dataset.map(
                combine_columns,
                fn_kwargs={
                    "text_columns": self.texts_column_names,
                    "image_columns": self.images_column_names,
                },
                desc="Combining columns",
            )
        else:
            dataset = self.dataset
            if isinstance(self.images_column_names, str) and self.images_column_names != "image":
                dataset = self.dataset.rename_column(
                    self.images_column_names, "image"
                )
            if isinstance(self.texts_column_names, str) and self.texts_column_names != "text":
                dataset = self.dataset.rename_column(
                    self.texts_column_names, "text"
                )

        dataset = prepare_image_dataset(dataset)

        num_images_per_sample = (
            len(self.images_column_names)
            if isinstance(self.images_column_names, list)
            else 1
        )
        num_texts_per_sample = (
            len(self.texts_column_names)
            if isinstance(self.texts_column_names, list)
            else 1
        )

        img_ground_truths = torch.arange(num_images_per_sample)
        caption_ground_truths = torch.arange(num_texts_per_sample)
        print("text encode")

        text_embeddings = model.encode(
            DataLoader(
                dataset.remove_columns("image"),
                batch_size=encode_kwargs["batch_size"],
            ),
            task_name=self.task_name,
        )
        print("image encode")
        image_embeddings = model.encode(
            DataLoader(
                dataset.remove_columns("text"),
                batch_size=encode_kwargs["batch_size"],
            ),
            task_name=self.task_name,
        )

        image_embeddings = F.normalize(
            image_embeddings,
            dim=-1,
        ).view(num_images_per_sample, -1)

        text_embeddings = F.normalize(
            text_embeddings,
            dim=-1,
        ).view(num_texts_per_sample, -1)

        image_score = []
        text_score = []
        score = []

        for img_emb, txt_emb in zip(image_embeddings, text_embeddings):
            scores = (
                img_emb @ txt_emb.t()
            )  # shape = (num_images_per_sample x num_texts_per_sample)

            image_closest_text = scores.argmax(
                dim=1
            )  # shape = (num_images_per_sample)
            text_closest_image = scores.argmax(
                dim=0
            )  # shape = (num_texts_per_sample)
            pred_text_is_correct = (
                (image_closest_text == img_ground_truths).all().item()
            )
            pred_image_is_correct = (
                (text_closest_image == caption_ground_truths).all().item()
            )
            all_correct = pred_text_is_correct and pred_image_is_correct
            image_score.append(pred_image_is_correct)
            text_score.append(pred_text_is_correct)
            score.append(all_correct)

        metrics = {}
        metrics["image_acc"] = torch.Tensor(image_score).float().mean().item()
        metrics["text_acc"] = torch.Tensor(text_score).float().mean().item()
        metrics["accuracy"] = torch.Tensor(score).float().mean().item()
        return metrics
```
</details>

**Comment by isaac-chung** [comment]
The current implementation uses a `ImageTextDataset`, which inherits from `torch.utils.data.Dataset`, which I think we agreed that it'd be acceptable as well. Also, if the test implementations are slower or use more memory, then perhaps we don't need to refactor it?

**Comment by Samoed** [comment]
Here is my results:
| task_name                 |   main_score (PR) |   main_score (main) |   eval_time (PR) |   eval_time (main) |
|:--------------------------|------------------:|--------------------:|-----------------:|-------------------:|
| AROFlickrOrder (ImageTextPairClassification)            |          0.5642   |            0.56     |         366.833  |           260.052  |
| BLINKIT2IMultiChoice      |          0.71144  |            0.71642  |          36.8783 |            37.9015 |
| BLINKIT2IRetrieval        |          0.47386  |            0.47478  |          41.9808 |            47.145  |
| BLINKIT2TMultiChoice      |          0.41362  |            0.41488  |          14.9453 |            24.2276 |
| BLINKIT2TRetrieval        |          0.3483   |            0.3492   |          18.7409 |            23.6017 |
| CIFAR10Clustering         |          0.692489 |            0.692489 |         131.455  |           125.352  |
| CIFAR10ZeroShot           |          0.9012   |            0.9012   |         113.235  |           121.646  |
| CUB200I2IRetrieval        |          0.61823  |            0.61892  |         349.657  |           174.404  |
| CVBenchCount (Any2TextMultichoice)              |          0.244924 |            0.244924 |          32.7175 |            82.6795 |
| Flickr30kT2IRetrieval     |          0.76779  |            0.7675   |          30.118  |            63.3901 |
| MNIST (classification)                    |          0.91108  |            0.91108  |         164.823  |           143.418  |
| MNISTZeroShot             |          0.6236   |            0.6236   |         112.496  |           114.574  |
| ROxfordEasyI2IMultiChoice |          0.0784   |            0.07923  |          74.0956 |            62.1927 |
| STS12VisualSTS            |        nan        |          nan        |         108.745  |           156.503  |
| Winoground (ImageTextPairClassification)               |          0.0775   |            0.0725   |         488.75   |            49.5028 |

Most of the time in `ImageTextPairClassification` is spent transforming images to RGB using `map` (~3 min), but I don't understand why it's so slow or how to improve it. Previously, I used `create_image_dataloader`, but I switched to transforming images with `map`. As a result, `AROFlickrOrder` initially took 147 seconds to evaluate, but after the change, it took 366 seconds. I made this change because some `Winoground` images have 4 dimensions instead of 3.

File: mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py [hunk 1]
```diff
@@ -63,22 +39,25 @@ def __call__(self, model: Encoder, *, encode_kwargs: dict[str, Any] = {}):
...
-        image_embeddings = model.get_image_embeddings(
-            dataloader, batch_size=encode_kwargs["batch_size"]
+        image_embeddings = model.encode(
+            dataloader,
+            task_name=self.task_name,
+            batch_size=encode_kwargs["batch_size"],
         )
 
+        # todo change to similarity
         probs = model.calculate_probs(text_embeddings, image_embeddings)
```
Comment (Samoed) [inline mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py:61]: Should we add `similarity` to the `Encoder` interface? MTEB encoders have similarity function, but MIEB have `calculate_probs`

**Review by KennethEnevoldsen — COMMENTED** [review]
Hmm, not too much of a variance, but it's a bit annoying that it is there. Do you have any idea why? Are we missing some controlled rng-states?

**Comment by Samoed** [comment]
I think there is some problems with RNG, I'll try to find what

**Comment by Samoed** [comment]
I tried to find what can cause differences in scores, but I can't find them

**Comment by isaac-chung** [comment]
Glad to see the times being faster. Were they all run on hot data?

**Comment by Samoed** [comment]
I've run all the tests with this code. The time I've spent on this is based on the JSON results.

```
@pytest.mark.parametrize(
    "task_name",
    [
        # "MNISTZeroShot",
        # "CIFAR10ZeroShot",
        # "MNIST",  # classification
        # "CIFAR10Clustering",
        # "STS12VisualSTS",
        # "CVBenchCount",  # any2text multichoice
        # "AROFlickrOrder",  # image text pair
        "CUB200I2IRetrieval",
        # "Winoground",  # image text pair
        # "ROxfordEasyI2IMultiChoice",  # i2i multichoice
        # "BLINKIT2IMultiChoice",  # it2i multichoice
        # "BLINKIT2TMultiChoice",  # it2t multichoice
        # "BLINKIT2IRetrieval",  # it2i retrieval
        # "BLINKIT2TRetrieval",  # it2t retrieval
        # "CUB200I2IRetrieval",  # i2i retrieval
        # "Flickr30kT2IRetrieval",  # t2i retrieval
    ],
)
def test_run_batched(task_name):
    task = mteb.get_task(task_name)
    model = mteb.get_model("openai/clip-vit-base-patch16")
    evaluator = mteb.MTEB([task])
    _ = evaluator.run(
        model,
        batch_size=32,
        output_folder="refactor_mieb_dataloaders/results",
        # output_folder="results",
        overwrite_results=True
    )
```

File: mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py (hunk 1 above)
Comment (isaac-chung) [reply mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py:61]: No strong opinions here. @KennethEnevoldsen or @gowitheflow-1998 ?

**Review by isaac-chung — APPROVED** [review]
Nice work!

File: mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py (hunk 1 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py:61]: Seems like a cosine similarity to me more than probabilities - would probably change

Comment (Samoed) [reply mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py:61]: Should I change `Encoder` interface to have similarity function?

**Comment by Samoed** [comment]
@isaac-chung @gowitheflow-1998 What should we do with `fusion_mode`? Most vision models have it, but we don’t save it in the results or metadata or use it directly in tasks. Maybe we should remove it, since different `fusion_mode` settings can lead to varying results.

File: mteb/models/vista_models.py [hunk 2]
```diff
@@ -100,123 +100,72 @@ def encode_text(self, texts):
...
-                        all_embeddings.append(batch_embeddings.cpu())
-            else:
-                assert len(texts) == len(images)
-                with torch.no_grad():
-                    for i in tqdm(range(0, len(texts), batch_size)):
-                        batch_texts = texts[i : i + batch_size]
-                        batch_images = images[i : i + batch_size]
-                        batch_embeddings = self.encode(
-                            images=batch_images, texts=batch_texts
+                        return self.encode_mm(
```
Comment (Samoed) [inline mteb/models/vista_models.py:158]: I don't understand clearrly how vista model work, because I can't find in implementation `encode_mm`

**Comment by Samoed** [comment]
@isaac-chung @KennethEnevoldsen I've tested models. Can you look?

File: mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py (hunk 1 above)
Comment (Samoed) [reply mteb/evaluation/evaluators/Image/ZeroshotClassificationEvaluator.py:61]: I think we can continue discussion in https://github.com/embeddings-benchmark/mteb/issues/1735

**Review by KennethEnevoldsen — APPROVED** [review]
I think this looks good. Couldn't find anything major. I would love to normalize the encoder interfaces as well, but I think that is for another PR
